Azure cloud provider: expose services on non-default subnets

itowlson opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**: The Azure cloud provider allows users to specify that a service should be exposed on an internal load balancer instead of the default external load balancer.  However, in a VNet environment, such services are currently always exposed on the master subnet.  Where there are multiple subnets in the VNet, it's desirable to be able to expose an internal service on any subnet.  This PR allows this via a new annotation, `service.beta.kubernetes.io/azure-load-balancer-internal-subnet`.

**Which issue this PR fixes**: fixes https://github.com/Azure/acs-engine/issues/1296 (no corresponding issue has been raised in the k8s core repo)

**Special notes for your reviewer**: None

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
A new service annotation has been added for services of type LoadBalancer on Azure, 
to specify the subnet on which the service's front end IP should be provisioned. The 
annotation is service.beta.kubernetes.io/azure-load-balancer-internal-subnet and its 
value is the subnet name (not the subnet ARM ID).  If omitted, the default is the 
master subnet.  It is ignored if the service is not on Azure, if the type is not 
LoadBalancer, or if the load balancer is not internal.
```

itowlson committed:
Azure: expose services on non-default subnets

itowlson commented:
/assign jdumars

jdumars commented:
/ok-to-test

itowlson commented:
/retest

itowlson commented:
/retest

itowlson commented:
/retest

itowlson commented:
/retest

jdumars commented:
/test pull-kubernetes-kubemark-e2e-gce

itowlson commented:
/test pull-kubernetes-kubemark-e2e-gce

karataliu commented on pkg/cloudprovider/providers/azure/azure_test.go:
```diff
@@ -887,3 +985,8 @@ func TestMetadataParsing(t *testing.T) {
 }
+
+func addTestSubnet(svc *v1.Service) {
+	svc.Annotations[ServiceAnnotationLoadBalancerInternal] = "true"
```
consider make use of getInternalTestService, and leave this func add subnet only.

karataliu commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -224,8 +230,17 @@ func serviceOwnsRule(service *v1.Service, rule string) bool {
 
-func getFrontendIPConfigName(service *v1.Service) string {
-	return cloudprovider.GetLoadBalancerName(service)
+func serviceOwnsFrontEndIP(fip network.FrontendIPConfiguration, service *v1.Service) bool {
```
Frontend

karataliu commented on pkg/cloudprovider/providers/azure/azure_test.go:
```diff
@@ -67,6 +67,104 @@ func TestReconcileLoadBalancerAddPort(t *testing.T) {
+	if err != nil {
+		t.Errorf("Unexpected error reconciling initial svc: %q", err)
+	}
+
```
Also validate FrontendIPConfigurations before editing subnet

karataliu commented on pkg/cloudprovider/providers/azure/azure_test.go:
```diff
@@ -67,6 +67,104 @@ func TestReconcileLoadBalancerAddPort(t *testing.T) {
+	az := getTestCloud()
+	svc := getTestService("servicea", v1.ProtocolTCP, 80)
+	addTestSubnet(&svc)
+	configProperties := getTestPublicFipConfigurationProperties()
```
This is logically incorrect, the property is for public IP case, though not validated in reconcileLoadBalancer.

itowlson commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -195,13 +195,19 @@ func getBackendPoolName(clusterName string) string {
 }
 
-func getSecurityRuleName(service *v1.Service, port v1.ServicePort, sourceAddrPrefix string) string {
+func getSecurityRuleName(service *v1.Service, port v1.ServicePort, subnetName *string, sourceAddrPrefix string) string {
```
You're right, this isn't needed.  I thought it was needed so that old security rules got replaced with new ones during an edit that changed the subnet, but looking at the properties, you're correct that we can just keep the existing rule.  Thanks!

itowlson commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -523,7 +531,7 @@ func (az *Cloud) reconcileLoadBalancer(lb network.LoadBalancer, fipConfiguration
 	lbName := getLoadBalancerName(clusterName, isInternal)
 	serviceName := getServiceName(service)
-	lbFrontendIPConfigName := getFrontendIPConfigName(service)
+	lbFrontendIPConfigName := getFrontendIPConfigName(service, subnet(service))
```
Fixed by making `subnet` check that the `Internal` annotation is also present.

itowlson committed:
Fixes issues noted in review

itowlson commented:
@karataliu Thanks for the detailed review - I appreciate your time. I think I've addressed the issues you've identified; please let me know if there are still outstanding problems.

itowlson commented:
/retest

karataliu commented on pkg/cloudprovider/providers/azure/azure_test.go:
```diff
@@ -67,6 +67,101 @@ func TestReconcileLoadBalancerAddPort(t *testing.T) {
 	validateLoadBalancer(t, lb, svc)
 }
 
+// Test additional of a new service/port on an internal LB with a subnet.
```
"a new service", since no port involved now.

karataliu reviewed (approved):
LGTM

itowlson commented:
/retest

brendandburns commented:
/lgtm
/approve no-isssue
